File: src/mission_controller/src/mission_controller/mission_controller.cpp
#include "mission_controller.h"
#include <pluginlib/class_list_macros.h>
#include <QStringList>

namespace mission_controller {


// Constructor: called before initPlugin function
MissionController::MissionController() : rqt_gui_cpp::Plugin(), widget_(0) {

    // Initialize QObject
    setObjectName("MissionController");

}

// 
void MissionController::initPlugin(qt_gui_cpp::PluginContext& context) {

    // Access standalone command line arguments
    QStringList argv context.argv();

    // Create QWidget
    widget_ = new QWidget();

    // Extend widget with all attributes and children from UI file
    ui_.setupUi(widget_);

    // Add widget to User Interface
    context.addWidget(widget_);

}

// Shuts down UI Plugin and unregisters all publishers
void MissionController::shutdownPlugin() {



}

// Save configuration
void MissionController::saveSettings(qt_gui_cpp::Settings& plugin_settings,
                                     const qt_gui_cpp::Settings& instance_settings) const {

// instance_settings.setValue(k,v)

} 

// Restore configuration
void MissionController::restoreSettings(const qt_gui_cpp::Settings& plugin_settings,
                                        const qt_gui_cpp::Settings& instance_settings) {

    // v = instance_settings.value(k)

} 

/*bool hasConfiguration() const {
  return true;
}

void triggerConfiguration() {
  // Usually used to open a dialog to offer the user a set of configurations
}*/

} // end namespace mission_controller

PLUGINLIB_DECLARE_CLASS(mission_controller, MissionController, 
                        mission_controller::MissionController, rqt_gui_cpp::MissionController)

File: src/mission_controller/src/mission_controller/mission_controller.py
#!/usr/bin/env python

from argparse import ArgumentParser
from qt_gui.plugin import Plugin
from .mission_controller_widget import MissionControllerWidget


class MissionController(Plugin):
    """ Subclass of Plugin that provides mission control options.
    """

    def __init__(self, context):
        """ Initializes widget and UI.
                :param context: plugin context hook to enable adding widgets as a ROS Gui pane, 'PluginContext'
        """

        # Get parent class
        super(MissionController, self).__init__(context)

        self.setObjectName('MissionController')

        self._widget = MissionControllerWidget(context)

        # For case when there are more than 1 of the widget in the rqt window
        if context.serial_number > 1:
            self._widget.setWindowTitle(self._widget.windowTitle() + (' (%d)' % context.serial_number()))
        
        # Add widgets to user interface
        context.add_widget(self._widget)

        # Parse command line arguments
        args = self.parse_args(context.argv())


    def parse_args(self, argv):
        """ Parses command line arguments. """

        parser = ArgumentParser(prog="mission_controller", add_help=False)
        MissionController.add_arguments(parser)
        return parser.parse_args(argv)


    @staticmethod
    def add_arguments(parser):
        """..."""

        group = parser.add_argument_group('Options for mission_controller plugin')
        #group.add_argument()


    def shutdown_plugin(self):
        """Unregisters all publishers from this widget."""

        pass


    def save_settings(self, plugin_settings, instance_settings):
        """Saves intrinsic plugin configuration."""

        pass


    def restore_settings(self, plugin_settings, instance_settings):
        """Restores intrinsic plugin configuration."""

        pass


    def trigger_configuration(self):
        """Opens settings dialog for plugin."""

        pass
